polaris/polaris/sep10auth/views.py

```diff
@@ -89,20 +90,114 @@ def _transaction_from_post_request(request):
     return _get_transaction_json(request.body)
 
 
-def _validate_envelope_xdr(envelope_xdr):
+def _validate_challenge_xdr(envelope_xdr):
     """
     Validate the provided TransactionEnvelope XDR (base64 string). Return the
     appropriate error if it fails, else the empty string.
+
+    verify_challenge_transaction() does not verify that the transaction's
+    signers are valid and have the total weight greater or equal to the
+    source account's medium threshold value, so this function does that as
+    well.
     """
     try:
         verify_challenge_transaction(
             challenge_transaction=envelope_xdr,
             server_account_id=settings.SIGNING_KEY,
             network_passphrase=settings.STELLAR_NETWORK_PASSPHRASE,
         )
-    except InvalidSep10ChallengeError as err:
-        return str(err)
-    return ""
+    except InvalidSep10ChallengeError as e:
+        raise ValueError(str(e))
+
+    transaction_envelope = TransactionEnvelope.from_xdr(
+        envelope_xdr, settings.STELLAR_NETWORK_PASSPHRASE
+    )
+    # verify_challenge_transaction ensures transaction has at least
+    # one operation whose source is the client account.
+    source_account = transaction_envelope.transaction.operations[0].source
+    tx_hash = transaction_envelope.hash()
+
+    # make horizon /accounts API call to get signers and threshold
+    account_signers, threshold = _get_signers_and_threshold(source_account)
+
+    ##################################################################
+    # The functionality implemented below should go in stellar_sdk and
+    # invoked like so:
+    #
+    #       verify_challenge_transaction(
+    #           challenge_transaction=envelope_xdr,
+    #           server_account_id=settings.SIGNING_KEY,
+    #           network_passphrase=settings.STELLAR_NETWORK_PASSPHRASE,
+    #           account_signers=account_signers,
+    #           threshold=threshold
```

Comment:
👏 I love this design of optionally providing the `account_signers` and the `threshold` on the existing function instead of making a new function. I'm guessing for the Python SDK we can make the behavior backwards compatible by allowing the `verify_challenge_transaction` function to operate without those fields and performing the existing behavior in that case. A client can call the function without the new parameters for accounts that do not exist yet just to validate the master key signed the transaction.

Reply:
Exactly.